File: built-in plugins dsp/mb-drums-percussion.cpp
/**
 * MB World Percussion
 * Category : instrument
 * Type     : drums
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Ethnic world percussion
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_DRUMS_PERCUSSION_H
#define MB_DRUMS_PERCUSSION_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbDrumsPercussion : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-drums-percussion";
    static constexpr const char* PLUGIN_NAME    = "MB World Percussion";
    static constexpr const char* PLUGIN_TYPE    = "drums";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float tone = 0.5f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbDrumsPercussion() = default;
    ~MbDrumsPercussion() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.tone = std::clamp(params.tone, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB World Percussion
        return input;
    }
};

#endif // MB_DRUMS_PERCUSSION_H

File: built-in plugins dsp/mb-drums-percussion.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbDrumsPercussionPlugin: PluginDefinition = {
    id: 'mb-drums-percussion', slug: 'mb-drums-percussion', name: 'MB World Percussion', category: 'instrument', type: 'drums', version: '1.0.0',
    description: 'Ethnic world percussion', author: 'Max Booster', grade: 'A',
    oscillators: [{ type: 'triangle', detune: 0, gain: 0.5 }, { type: 'noise', detune: 0, gain: 0.3 }, { type: 'sine', detune: 0, gain: 0.4 }],
    envelope: { attack: 0.002, decay: 0.2, sustain: 0.0, release: 0.25 },
    parameters: [
      { id: 'tone', name: 'Tone', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true },
      { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { tone: 0.5, volume: 0.8 },
  };

export default MbDrumsPercussionPlugin;
